### PR TITLE
k8s - patch existing resource

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

add new value ``patched`` for  ``state`` option for ``k8s`` module. patched existing resource with new ``definition``

#84 
##### ISSUE TYPE

- Feature Pull Request
##### COMPONENT NAME

k8s
##### ADDITIONAL INFORMATION

```paste below
kubernetes.core.k8s:
  state: patched
  wait: yes
  definition: |
    ---
    apiVersion: v1
    kind: Namespace
    metadata:
      name: "{{ patch_only_namespace.first }}"
      labels:
        patch: ansible
    ---
    apiVersion: v1
    kind: Namespace
    metadata:
      name: "{{ patch_only_namespace.second }}"
      labels:
        patch: ansible
```

### CODE REVIEW

**Comment by gravesm** [comment]
This is probably the cleanest, easiest solution to the immediate problem. I can't think of a better one, but I also have some concerns.

My first concern is that this feels like it kind of goes against how Ansible usually operates. We're using `state: present` which would typically mean creating a thing, but then providing a second parameter to essentially override that. I think this makes the module somewhat harder to use correctly and could lead to more frequent user error.

My second concern is that while this is backwards compatible, which is great, the default behavior of the module is still broken. A patch is designed to modify an existing resource. IMO if the user has supplied a patch our default should be to not create the resource.

**Comment by tima** [comment]
Well reasoned @gravesm. 

To the first concern, I agree while it works it does make the module interface less elegant and more difficult to use. Is there some "prior art" example in the Ansible module space of a state that is "make a change, but don't create it" we could implement here? So I agree. Words matter and I am looking for the right word for this state.

To the second concern, I see what you mean. This is a tough one in that the default behavior is indeed broken so do we fix that behavior and potentially break users who expect it to work that way OR do we let it continue as is and document the inconsistency? I suppose a middle ground is to call it deprecated behavior and emit a warning that they need to switch to what ever we come up with.

**Comment by tima** [comment]
I was browsing modules looking for one that has something that isn't just present and absent. I came across [ansible.windows.win_updates](https://docs.ansible.com/ansible/latest/collections/ansible/windows/win_updates_module.html) that has "installed" as a state. While you don't install a resource, you do defined one. So would `state: defined` be acceptable? Too confusing to separate from absent? 

I haven't had a lot of luck finding alternatives so far otherwise.

**Comment by gravesm** [comment]
A couple other examples with more granular state parameters:
* `file`: absent, directory, file, hard, link, touch
* `apt`: absent, build-dep, latest, present, fixed
* `dnf`: absent, present, installed, removed, latest

I think there's definitely prior art here. If we wanted to maintain backwards compatibility we could perhaps add a new state called `patched` that would only modify existing resources?

**Comment by Akasurde** [comment]
@gravesm I totally agree with you. We can have an additional parameter like `action=patched` with `state=present` so that we align with Ansible Standards. vmware_rest uses this kind of thing where we have `state` different than `absent`, `present`.

**Comment by gravesm** [comment]
@Akasurde could you clarify? Are you suggesting we should add a new state `patched` or are you suggesting we should add a new parameter `action=patched` with `state=present`.

**Review by gravesm — APPROVED** [review]

**Comment by Akasurde** [comment]
> @Akasurde could you clarify? Are you suggesting we should add a new state `patched` or are you suggesting we should add a new parameter `action=patched` with `state=present`.

we should add a new parameter `action=patched` with `state=present` if everyone agrees.

**Comment by abikouo** [comment]
> > @Akasurde could you clarify? Are you suggesting we should add a new state `patched` or are you suggesting we should add a new parameter `action=patched` with `state=present`.
> 
> we should add a new parameter `action=patched` with `state=present` if everyone agrees.

we are increasing the complexity of this module, we should rather go for something different, for instance, a new module dealing with all  patching functionalities

**Comment by Akasurde** [comment]
I am OK with new module as well.

**Comment by jmazzitelli** [comment]
`state=patched` seems fine to me. With that set, the context should be clear - if the object doesn't exist, it won't be created nor will that be considered an error condition. It's just "patch an existing resource". That seems clear to me from an end-user (assuming that IS what we are talking about - if it is not, then it is not clear :))

File: molecule/default/tasks/patched.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+          first: patched-namespace-1
+          second: patched-namespace-2
+
+    - name: Ensure namespace {{ patch_only_namespace.first }} exist
+      kubernetes.core.k8s:
+        definition:
+          apiVersion: v1
+          kind: Namespace
+          metadata:
+            name: "{{ patch_only_namespace.first }}"
```
Comment (jmazzitelli) [inline molecule/default/tasks/patched.yml:14]: Could I ask that this test check to ensure that other parts of the resource are not touched (i.e. left as-is). For example, adding a label and annotation and then later testing that that label and annotation still exist after the patch:

```suggestion
            name: "{{ patch_only_namespace.first }}"
            labels:
              existingLabel: "labelValue"
            annotations:
              existingAnnotation: "annotationValue"
```

File: molecule/default/tasks/patched.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            labels:
+              ansible: patched
+      register: patch_resource
+
+    - name: assert that patch succeed
+      assert:
+        that:
+          - patch_resource.changed
+          - patch_resource.result.results | selectattr('warning', 'defined') | list | length == 1
+
```
Comment (jmazzitelli) [inline molecule/default/tasks/patched.yml:54]: This code might not be exactly right - but hopefully you get the idea. Get the first namespace and make sure the labels and annotations are as expected (you should see the old label and old annotation as well as the new label).

```suggestion

    - name: Ensure namespace {{ patch_only_namespace.first }} was patched correctly
      kubernetes.core.k8s_info:
        kind: namespace
        name: "{{ patch_only_namespace.first }}"
      register: first_namespace

    - name: assert labels are as expected
      assert:
        that:
          - first_namespace.resources[0].metadata.labels.ansible == "patched"
          - first_namespace.resources[0].metadata.labels.existingLabel == "labelValue"
          - first_namespace.resources[0].metadata.annotations.existingAnnotation == "annotationValue"
```

**Review by jmazzitelli — COMMENTED** [review]
I would suggest adding some tests to make sure the patch does what was expected (check to make sure existing data remains intact (e.g. existing labels and annotations) and see the new label was added).

**Pull request merged**